jetstream/engine/aot_utils.py

```diff
@@ -0,0 +1,246 @@
+
+
+def initialize_insert_generate_jit_cache(
+    *,
+    prefill_engine: engine_api.WarmedUpEngine,
+    generate_engine: engine_api.WarmedUpEngine,
+    prefill_params: Any,
+    generate_params: Any,
+    generate_idx: int,
+):
+  """Initialiszes jit cache for insert and generate.
+
+  Args:
+      generate_engine: A generate engine to be compiled for.
+      generate_params: The associated parameters.
+      generate_idx: Which generate engine it is.
+  """
+
+  prefill_buckets = token_utils.DEFAULT_PREFILL_BUCKETS
+  prefill_buckets = [
+      bucket
+      for bucket in prefill_buckets
+      if bucket <= generate_engine.max_prefill_length
+  ]
+  generate_engine.prefill_buckets = prefill_buckets
+  if generate_engine.max_prefill_length not in prefill_buckets:
+    prefill_buckets.append(generate_engine.max_prefill_length)
+
+  decode_state = generate_engine.init_decode_state()
+
+  def compile_insert(length):
+    padded_tokens, true_length = jnp.ones((length), dtype="int32"), length
+
+    prefill, _ = prefill_engine._downstream_engine.prefill(  # pylint: disable=protected-access
+        params=prefill_params,
+        padded_tokens=padded_tokens,
+        true_length=true_length,
+    )
+
+    lowered = jax.jit(generate_engine._downstream_engine.insert).lower(  # pylint: disable=protected-access
+        prefix=prefill, decode_state=decode_state, slot=1
+    )
+    logging.info(
+        "---------Generate engine %d lowered for insert length %d.---------",
+        generate_idx,
+        length,
+    )
+    compiled = lowered.compile()
+    insert_compiled[length] = compiled
+
+    logging.info(
+        "---------Generate engine %d compiled for insert length %d.---------",
+        generate_idx,
+        length,
+    )
+
+  def compile_generate():
+
+    logging.info(
+        "---------Generate compilation %d begun.---------", generate_idx
+    )
+
+    lowered = jax.jit(generate_engine._downstream_engine.generate).lower(  # pylint: disable=protected-access
+        params=generate_params,
+        decode_state=decode_state,
+    )
+    logging.info(
+        "---------Generate engine %d lowered.---------",
+        generate_idx,
+    )
+
+    compiled = lowered.compile()
+    logging.info(
+        "---------Generate engine %d compiled.---------",
+        generate_idx,
+    )
+
+    logging.info(
+        "---------Generate compilation %d complete.---------", generate_idx
+    )
+
+    return compiled
+
+  logging.info(
+      "---------Insertion generation compilation %d begun.---------",
+      generate_idx,
+  )
+
+  generate_compiled = compile_generate()
+  logging.info(
+      "---------Generate engine %d compiled generation step.---------",
+      generate_idx,
+  )
+  generate_engine.generate_compiled = generate_compiled
+
+  insert_compiled = {}
+  with concurrent.futures.ThreadPoolExecutor(
+      max_workers=len(prefill_buckets)
+  ) as executor:
+    _ = list(executor.map(compile_insert, prefill_buckets))
```

Comment:
insert compiled is not updated here?

Reply:
It is updated in the compile_insert, but I changed it to be updated outside now!